File: apps/api/presentation/restapi/checklist_session_transformer.go
package restapi

import (
	"apps/api/domain"
	"encoding/json"
	apiContract "libs/api-contract"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

func GetChecklistSessionId(r *http.Request) (int64, error) {
	vars := mux.Vars(r)
	idParam := vars["checklistSessionId"]
	id, err := strconv.ParseInt(idParam, 10, 64)
	return id, err
}

func GetChecklistSessionItemId(r *http.Request) (int64, error) {
	vars := mux.Vars(r)
	idParam := vars["checklistSessionItemId"]
	id, err := strconv.ParseInt(idParam, 10, 64)
	return id, err
}

func GetChecklistSessionSubItemId(r *http.Request) (int64, error) {
	vars := mux.Vars(r)
	idParam := vars["checklistSessionSubItemId"]
	id, err := strconv.ParseInt(idParam, 10, 64)
	return id, err
}

func GetChecklistSessionRequest(r *http.Request) (apiContract.ChecklistSessionRequest, error) {
	var req apiContract.ChecklistSessionRequest
	err := json.NewDecoder(r.Body).Decode(&req)
	return req, err
}

func ToApiChecklistSessionSubItem(d domain.ChecklistSessionSubItem) apiContract.ChecklistSessionSubItem {
	return apiContract.ChecklistSessionSubItem{
		Id:                         d.Id,
		ChecklistSessionItemId:     d.ChecklistSessionItemId,
		ChecklistTemplateSubItemId: d.ChecklistTemplateSubItemId,
		Name:                       d.Name,
		DisplayOrder:               d.DisplayOrder,
		CompletedAt:                d.CompletedAt,
		CreatedAt:                  d.CreatedAt,
		UpdatedAt:                  d.UpdatedAt,
	}
}

func ToApiChecklistSessionItem(d domain.ChecklistSessionItem) apiContract.ChecklistSessionItem {
	subItems := []apiContract.ChecklistSessionSubItem{}
	for _, si := range d.SubItems {
		subItems = append(subItems, ToApiChecklistSessionSubItem(si))
	}
	return apiContract.ChecklistSessionItem{
		Id:                      d.Id,
		ChecklistSessionId:      d.ChecklistSessionId,
		ChecklistTemplateItemId: d.ChecklistTemplateItemId,
		Name:                    d.Name,
		Description:             d.Description,
		DisplayOrder:            d.DisplayOrder,
		CompletedAt:             d.CompletedAt,
		SubItems:                subItems,
		CreatedAt:               d.CreatedAt,
		UpdatedAt:               d.UpdatedAt,
	}
}

func ToApiChecklistSession(d domain.ChecklistSession) apiContract.ChecklistSession {
	items := []apiContract.ChecklistSessionItem{}
	for _, item := range d.Items {
		items = append(items, ToApiChecklistSessionItem(item))
	}

	var apiTemplate *apiContract.ChecklistTemplate
	if d.ChecklistTemplate != nil {
		t := ToApiChecklistTemplate(*d.ChecklistTemplate)
		apiTemplate = &t
	}

	return apiContract.ChecklistSession{
		Id:                  d.Id,
		ChecklistTemplateId: d.ChecklistTemplateId,
		ChecklistTemplate:   apiTemplate,
		Date:                d.Date,
		CompletedAt:         d.CompletedAt,
		Items:               items,
		CreatedAt:           d.CreatedAt,
		UpdatedAt:           d.UpdatedAt,
		DeletedAt:           d.DeletedAt,
	}
}

func ToChecklistSession(req apiContract.ChecklistSessionRequest) domain.ChecklistSession {
	return domain.ChecklistSession{
		ChecklistTemplateId: req.ChecklistTemplateId,
		Date:                req.Date,
	}
}

File: libs/ui/src/data/api/checklistSession.transformer.ts
// eslint-disable-next-line @nx/enforce-module-boundaries
import {
  ChecklistSession as ApiChecklistSession,
  ChecklistSessionItem as ApiChecklistSessionItem,
  ChecklistSessionSubItem as ApiChecklistSessionSubItem,
  ChecklistSessionRequest,
} from '../../../../api-contract/src';
import {
  ChecklistSession,
  ChecklistSessionForm,
  ChecklistSessionItem,
  ChecklistSessionSubItem,
} from '../../domain';
import { toChecklistTemplate } from './checklistTemplate.transformer';

export function toChecklistSessionSubItem(
  sub: ApiChecklistSessionSubItem
): ChecklistSessionSubItem {
  return {
    id: sub.id,
    checklistSessionItemId: sub.checklistSessionItemId,
    checklistTemplateSubItemId: sub.checklistTemplateSubItemId ?? null,
    name: sub.name,
    displayOrder: sub.displayOrder,
    completedAt: sub.completedAt ?? null,
    createdAt: sub.createdAt,
    updatedAt: sub.updatedAt,
  };
}

export function toChecklistSessionItem(
  item: ApiChecklistSessionItem
): ChecklistSessionItem {
  return {
    id: item.id,
    checklistSessionId: item.checklistSessionId,
    checklistTemplateItemId: item.checklistTemplateItemId ?? null,
    name: item.name,
    description: item.description ?? null,
    displayOrder: item.displayOrder,
    completedAt: item.completedAt ?? null,
    subItems: item.subItems.map(toChecklistSessionSubItem),
    createdAt: item.createdAt,
    updatedAt: item.updatedAt,
  };
}

export function toChecklistSession(s: ApiChecklistSession): ChecklistSession {
  return {
    id: s.id,
    checklistTemplateId: s.checklistTemplateId,
    checklistTemplate: s.checklistTemplate
      ? toChecklistTemplate(s.checklistTemplate)
      : null,
    date: s.date,
    completedAt: s.completedAt ?? null,
    items: s.items.map(toChecklistSessionItem),
    createdAt: s.createdAt,
    updatedAt: s.updatedAt,
  };
}

export function toApiChecklistSession(
  form: ChecklistSessionForm
): ChecklistSessionRequest {
  return {
    checklistTemplateId: form.checklistTemplateId,
    date: form.date,
  };
}
